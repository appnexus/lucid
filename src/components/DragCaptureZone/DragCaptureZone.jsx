import _ from 'lodash';
import React from 'react';
import { lucidClassNames } from '../../util/style-helpers';
import { createClass } from '../../util/component-definition';

const boundClassNames = lucidClassNames.bind('&-DragCaptureZone');
const {
	func,
	string
} = React.PropTypes;

/**
 * {"categories": ["utility"]}
 *
 * This is a helper component used to capture mouse events to determine when the
 * user starts, is and stops dragging.
 */
const DragCaptureZone = createClass({
	displayName: 'DragCaptureZone',
	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root
		 * element.
		 */
		className: string,

		/**
		 * Called as the user drags the mouse.
		 *
		 * Signature: `({ dx, dy, pageX, pageY }, { event, props }) => {}`
		 */
		onDrag: func,

		/**
		 * Called when the user releases the mouse button after having dragged.
		 *
		 * Signature: `({ dx, dy, pageX, pageY }, { event, props }) => {}`
		 */
		onDragEnd: func,

		/**
		 * Called when the user presses the mouse button down while over the
		 * component.
		 *
		 * Signature: `({ dx, dy, pageX, pageY }, { event, props }) => {}`
		 */
		onDragStart: func
	},

	getDefaultProps() {
		return {
			onDrag: _.noop,
			onDragEnd: _.noop,
			onDragStart: _.noop
		};
	},

	getInitialState() {
		return {
			pageX: 0,
			pageY: 0
		};
	},

	render() {
		return (
			<div
				{...this.props}
				className={boundClassNames('&', this.props.className)}
				key='DragCaptureZone'
				onMouseDown={this.handleDragStart}
			/>
		);
	},

	componentWillUnmount() {
		window.document.removeEventListener('mousemove', this.handleDrag);
		window.document.removeEventListener('mouseup', this.handleDragEnd);
	},

	handleDrag(event) {
		const {
			pageX,
			pageY
		} = event;

		event.preventDefault();

		this.props.onDrag({
			dX: pageX - this.state.pageX,
			dY: pageY - this.state.pageY,
			pageX,
			pageY
		}, {
			event,
			props: this.props
		});
	},

	handleDragEnd(event) {
		const {
			pageX,
			pageY
		} = event;

		event.preventDefault();

		window.document.removeEventListener('mousemove', this.handleDrag);
		window.document.removeEventListener('mouseup', this.handleDragEnd);

		this.props.onDragEnd({
			dX: pageX - this.state.pageX,
			dY: pageY - this.state.pageY,
			pageX,
			pageY
		}, {
			event,
			props: this.props
		});

		this.setState({
			pageX: 0,
			pageY: 0
		});
	},

	handleDragStart(event) {
		const {
			pageX,
			pageY
		} = event;

		event.preventDefault();

		window.document.addEventListener('mousemove', this.handleDrag);
		window.document.addEventListener('mouseup', this.handleDragEnd);

		this.props.onDragStart({
			dX: 0,
			dY: 0,
			pageX,
			pageY
		}, {
			event,
			props: this.props
		});

		this.setState({
			pageX,
			pageY
		});
	}
});

export default DragCaptureZone;
