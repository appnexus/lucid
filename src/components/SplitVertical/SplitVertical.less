@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

.lucid-SplitVertical {

	& > &-inner {
		& > .lucid-SplitVertical-Divider {

			background: @color-borderColor;
			box-sizing: border-box;
			width: 1px;

			&.lucid-SplitVertical-Divider-is-resizeable {
				background: @color-gray;
				cursor: ew-resize;
			}
		}
	}
}
