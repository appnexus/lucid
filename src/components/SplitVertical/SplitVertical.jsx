import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { lucidClassNames } from '../../util/style-helpers';
import {
	createClass as createReactClass,
	filterTypes,
	omitProps,
} from '../../util/component-types';
import DragCaptureZone from '../DragCaptureZone/DragCaptureZone';
import { Motion, spring } from 'react-motion';
import { QUICK_SLIDE_MOTION } from '../../constants/motion-spring';

const cx = lucidClassNames.bind('&-SplitVertical');

const { any, bool, func, node, number, string, oneOfType } = PropTypes;

/**
 * {"categories": ["helpers"]}
 *
 * `SplitVertical` renders a vertical split.
 */
const SplitVertical = createReactClass({
	displayName: 'SplitVertical',

	_isPrivate: true,

	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root
		 * element. Value is run through the `classnames` library.
		 */
		className: any,
		/**
		 * Direct children must be types {Splitvertical.Leftpane, Splitvertical.Divider, Splitvertical.RightPane}.
		 * All content is composed as children of these respective elements.
		 */
		children: node,
		/**
		 * Allows draggable resizing of the SplitVertical
		 */
		isResizeable: bool,
		/**
		 * Render as expanded or collapsed.
		 */
		isExpanded: bool,
		/**
		 * Allows animated expand and collapse behavior.
		 */
		isAnimated: bool,
		/**
		 * Called when the user is currently resizing the split with the Divider.
		 *
		 * Signature: `(width, { event, props }) => {}`
		 */
		onResizing: func,
		/**
		 * Called when the user resizes the split with the Divider.
		 *
		 * Signature: `(width, { event, props }) => {}`
		 */
		onResize: func,
		/**
		 * Use this prop to shift the collapsed position by a known value.
		 */
		collapseShift: number,
	},

	components: {
		/**
		 * Left pane of the split.
		 */
		LeftPane: createReactClass({
			displayName: 'SplitVertical.LeftPane',
			propTypes: {
				/**
				 * Any valid React children.
				 */
				children: node,
				/**
				 * Set width of this pane.
				 */
				width: oneOfType([number, string]),
				/**
				 * Define this pane as the primary content pane. When the split is collapsed, this pane becomes full width.
				 */
				isPrimary: bool,
			},
			getDefaultProps() {
				return {
					isPrimary: false,
				};
			},
		}),

		/**
		 * Right pane of the split.
		 */
		RightPane: createReactClass({
			displayName: 'SplitVertical.RightPane',
			propTypes: {
				/**
				 * Any valid React children.
				 */
				children: node,
				/**
				 * Set width of this pane.
				 */
				width: oneOfType([number, string]),
				/**
				 * Define this pane as the primary content pane. When the split is collapsed, this pane becomes full width.
				 */
				isPrimary: bool,
			},
			getDefaultProps() {
				return {
					isPrimary: false,
				};
			},
		}),

		/**
		 * The area that separates the split panes. Can be dragged to resize them.
		 */
		Divider: createReactClass({
			displayName: 'SplitVertical.Divider',
			propTypes: {
				/**
				 * Any valid React children.
				 */
				children: node,
			},
		}),
	},

	getDefaultProps() {
		return {
			isExpanded: true,
			isAnimated: false,
			collapseShift: 0,
			onResizing: _.noop,
			onResize: _.noop,
			isResizeable: true,
		};
	},

	getInitialState() {
		return {
			isAnimated: false, // to ensure first render doesn't show a collapse animation
			isExpanded: true,
			collapseAmount: 250,
		};
	},

	getPanes() {
		const { children } = this.props;
		const { leftPane: leftPaneRef, rightPane: rightPaneRef } = this.storedRefs;

		const leftPaneElement = _.get(
			filterTypes(children, SplitVertical.LeftPane),
			0,
			<SplitVertical.LeftPane />
		);
		const rightPaneElement = _.get(
			filterTypes(children, SplitVertical.RightPane),
			0,
			<SplitVertical.RightPane />
		);
		let primaryElement, primaryRef;
		let secondaryElement, secondaryRef;

		if (leftPaneElement.props.isPrimary && !rightPaneElement.props.isPrimary) {
			primaryElement = leftPaneElement;
			primaryRef = leftPaneRef;
			secondaryElement = rightPaneElement;
			secondaryRef = rightPaneRef;
		} else {
			primaryElement = rightPaneElement;
			primaryRef = rightPaneRef;
			secondaryElement = leftPaneElement;
			secondaryRef = leftPaneRef;
		}

		return {
			left: leftPaneElement.props,
			right: rightPaneElement.props,
			primary: primaryElement.props,
			primaryRef,
			secondary: secondaryElement.props,
			secondaryRef,
		};
	},

	// Style changes to DOM nodes are updated here to shortcut the state -> render cycle for better performance. Also the Style updates in this
	// function are entirely transient and can be flushed with a props update to `width`.
	applyDeltaToSecondaryWidth(
		dX,
		isExpanded,
		secondaryStartRect,
		secondaryRef,
		secondary,
		right,
		innerRef,
		primaryRef,
		collapseShift = 0
	) {
		if (isExpanded) {
			secondaryRef.style.flexBasis = `${secondaryStartRect.width + dX * (secondary === right ? -1 : 1)}px`;
			return secondaryStartRect.width + dX * (secondary === right ? -1 : 1);
		} else {
			const overlapWidth =
				(secondary === right
					? secondaryStartRect.width + dX
					: secondaryStartRect.width - dX) - collapseShift;

			if (overlapWidth > 0) {
				this.collapseSecondary(overlapWidth);
				return secondaryStartRect.width - overlapWidth;
			} else {
				this.expandSecondary();
				secondaryRef.style.flexBasis = `${(dX + collapseShift) * (secondary === right ? -1 : 1)}px`;
				return (dX + collapseShift) * (secondary === right ? -1 : 1);
			}
		}
	},

	expandSecondary() {
		this.setState({ isExpanded: true });
	},

	collapseSecondary(collapseAmount) {
		this.setState({ isExpanded: false, collapseAmount });
	},

	disableAnimation(innerRef, secondaryRef, primaryRef) {
		innerRef.style.transition = 'all 0s';
		secondaryRef.style.transition = 'all 0s';
		primaryRef.style.transition = 'all 0s';
	},

	resetAnimation(innerRef, secondaryRef, primaryRef) {
		innerRef.style.transition = '';
		secondaryRef.style.transition = '';
		primaryRef.style.transition = '';
	},

	handleDragStart() {
		this.panes = this.getPanes();
		const { secondaryRef, primaryRef } = this.panes;
		this.secondaryStartRect = secondaryRef.getBoundingClientRect();
		this.disableAnimation(this.storedRefs.inner, secondaryRef, primaryRef);
	},

	handleDrag({ dX }, { event }) {
		const { isExpanded, collapseShift, onResizing } = this.props;

		const { secondaryRef, secondary, right, primaryRef } = this.panes;

		onResizing(
			this.applyDeltaToSecondaryWidth(
				dX,
				isExpanded,
				this.secondaryStartRect,
				secondaryRef,
				secondary,
				right,
				this.storedRefs.inner,
				primaryRef,
				collapseShift
			),
			{ props: this.props, event }
		);
	},

	handleDragEnd({ dX }, { event }) {
		const { isExpanded, collapseShift, onResize } = this.props;

		const { secondaryRef, secondary, right, primaryRef } = this.panes;

		onResize(
			this.applyDeltaToSecondaryWidth(
				dX,
				isExpanded,
				this.secondaryStartRect,
				secondaryRef,
				secondary,
				right,
				this.storedRefs.inner,
				primaryRef,
				collapseShift
			),
			{ props: this.props, event }
		);

		this.resetAnimation(this.storedRefs.inner, secondaryRef, primaryRef);
	},

	componentWillReceiveProps(nextProps) {
		const { isAnimated, isExpanded, collapseShift } = nextProps;

		const { secondaryRef } = this.getPanes();

		if (
			!isExpanded && // check if collapseShift changed or secondary pane collapsed
			(this.props.isExpanded || this.props.collapseShift !== collapseShift)
		) {
			// collapse secondary
			const secondaryRect = secondaryRef.getBoundingClientRect();
			this.collapseSecondary(secondaryRect.width - collapseShift);
		} else if (!this.props.isExpanded && isExpanded) {
			// expand secondary
			this.expandSecondary();
		}

		if (this.state.isAnimated !== isAnimated) {
			this.setState({
				isAnimated,
			});
		}
	},

	componentDidMount() {
		const { isExpanded, isAnimated, collapseShift } = this.props;

		const { primaryRef, secondaryRef } = this.getPanes();

		const { inner } = this.storedRefs;

		_.defer(() => {
			this.disableAnimation(inner, secondaryRef, primaryRef);
			inner.style.visibility = 'hidden';
			_.defer(() => {
				if (!isExpanded) {
					// collapse secondary
					const secondaryRect = secondaryRef.getBoundingClientRect();
					this.collapseSecondary(secondaryRect.width - collapseShift);
				}

				_.defer(() => {
					if (isAnimated) {
						this.setState({ isAnimated });
					}
					_.defer(() => {
						inner.style.visibility = '';
						this.resetAnimation(inner, secondaryRef, primaryRef);
					});
				});
			});
		});
	},

	storeRef(name) {
		return ref => {
			this.storedRefs[name] = ref;
		};
	},

	componentWillMount() {
		this.storedRefs = {};
	},

	render() {
		const { children, className, isResizeable, ...passThroughs } = this.props;

		const { isAnimated, isExpanded, collapseAmount } = this.state;

		const {
			left: leftPaneProps,
			right: rightPaneProps,
			secondary,
		} = this.getPanes();

		const dividerProps = _.get(
			_.first(filterTypes(children, SplitVertical.Divider)),
			'props',
			{}
		);

		let from, to;

		if (!isExpanded) {
			from = { slideAmount: 0 };
			to = { slideAmount: collapseAmount };
		} else {
			from = { slideAmount: 0 };
			to = { slideAmount: 0 };
		}

		const isRightSecondary = rightPaneProps === secondary;

		return (
			<div
				{...omitProps(passThroughs, SplitVertical)}
				className={cx(
					'&',
					{
						'&-is-expanded': isExpanded,
						'&-is-animated': isAnimated,
					},
					className
				)}
				style={{
					overflow: 'hidden',
					...passThroughs.style,
				}}
			>
				<Motion
					defaultStyle={from}
					style={
						isAnimated
							? _.mapValues(to, val => spring(val, QUICK_SLIDE_MOTION))
							: to
					}
				>
					{tween => (
						<div
							className={cx('&-inner')}
							ref={this.storeRef('inner')}
							style={{
								display: 'flex',
								transform: `translateX(${(isRightSecondary ? 1 : -1) * Math.round(tween.slideAmount)}px)`,
							}}
						>
							<div
								{...omitProps(leftPaneProps, SplitVertical.LeftPane)}
								className={cx(
									'&-LeftPane',
									{
										'&-is-secondary': leftPaneProps === secondary,
									},
									leftPaneProps.className
								)}
								style={{
									flexGrow: isRightSecondary ? 1 : 0,
									flexShrink: isRightSecondary ? 1 : 0,
									flexBasis: _.isNil(leftPaneProps.width)
										? leftPaneProps === secondary ? 'calc(50% - 3px)' : '0%'
										: leftPaneProps.width,
									marginLeft: isRightSecondary
										? -Math.round(tween.slideAmount)
										: null,
									overflow: 'auto',
									...leftPaneProps.style,
								}}
								ref={this.storeRef('leftPane')}
							>
								{leftPaneProps.children}
							</div>
							{isResizeable
								? <DragCaptureZone
										{...omitProps(
											dividerProps,
											SplitVertical.Divider,
											[],
											false
										)}
										className={cx(
											'&-Divider',
											'&-Divider-is-resizeable',
											dividerProps.className
										)}
										onDragStart={this.handleDragStart}
										onDrag={this.handleDrag}
										onDragEnd={this.handleDragEnd}
										style={{
											width: '6px',
											boxSizing: 'border-box',
											...dividerProps.style,
										}}
									>
										{dividerProps.children || ' '}
									</DragCaptureZone>
								: <div
										{...omitProps(dividerProps, SplitVertical.Divider)}
										className={cx('&-Divider', dividerProps.className)}
									>
										{dividerProps.children || ' '}
									</div>}
							<div
								{...omitProps(rightPaneProps, SplitVertical.RightPane)}
								className={cx(
									'&-RightPane',
									{
										'&-is-secondary': rightPaneProps === secondary,
									},
									rightPaneProps.className
								)}
								style={{
									flexGrow: !isRightSecondary ? 1 : 0,
									flexShrink: !isRightSecondary ? 1 : 0,
									flexBasis: _.isNil(rightPaneProps.width)
										? rightPaneProps === secondary ? 'calc(50% - 3px)' : '0%'
										: rightPaneProps.width,
									marginRight: isRightSecondary
										? null
										: -Math.round(tween.slideAmount),
									overflow: 'auto',
									...rightPaneProps.style,
								}}
								ref={this.storeRef('rightPane')}
							>
								{rightPaneProps.children}
							</div>
						</div>
					)}
				</Motion>
			</div>
		);
	},
});

export default SplitVertical;
