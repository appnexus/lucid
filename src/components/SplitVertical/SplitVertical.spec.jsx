import React from 'react';
import { shallow, mount } from 'enzyme';
import assert from 'assert';
import sinon from 'sinon';
import describeWithDOM from '../../util/describe-with-dom';
import _ from 'lodash';
import { common } from '../../util/generic-tests';
import SplitVertical from './SplitVertical';
import DragCaptureZone from '../DragCaptureZone/DragCaptureZone';

describe('SplitVertical', () => {
	common(SplitVertical);

	describe('render', () => {
		it('should render an inner element with the left, right, and divider elements as children', () => {
			const wrapper = shallow(
				<SplitVertical />
			);

			assert.equal(wrapper.find('.lucid-SplitVertical').length, 1);
			assert.equal(wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner').length, 1);
			assert.equal(wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-LeftPane').length, 1);
			assert.equal(wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-Divider').length, 1);
			assert.equal(wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-RightPane').length, 1);
		});
	});

	describe('props', () => {
		describe('isExpanded', () => {
			it('should default to true', () => {
				const wrapper = shallow(
					<SplitVertical />
				);

				assert.equal(wrapper.find('.lucid-SplitVertical.lucid-SplitVertical-is-expanded').length, 1);
			});

			it('should apply the &-is-expanded css class when true', () => {
				const wrapper = shallow(
					<SplitVertical isExpanded={true} />
				);

				assert.equal(wrapper.find('.lucid-SplitVertical.lucid-SplitVertical-is-expanded').length, 1);
			});

			it('should apply the &-is-expanded css class when false', () => {
				const wrapper = shallow(
					<SplitVertical isExpanded={false} />
				);

				assert.equal(wrapper.find('.lucid-SplitVertical.lucid-SplitVertical-is-expanded').length, 0);
			});
		});

		describeWithDOM('isAnimated', () => {
			let wrapper;

			afterEach(() => {
				if (wrapper) {
					wrapper.unmount();
				}
			});

			it('should default to false', (done) => {
				wrapper = mount(
					<SplitVertical />
				);

				_.delay(() => {
					assert.equal(wrapper.find('.lucid-SplitVertical.lucid-SplitVertical-is-animated').length, 0);
					done();
				}, 10)
			});

			it('should apply the &-is-animated class when true, after initial render', (done) => {
				wrapper = mount(
					<SplitVertical isAnimated={true} />
				);

				_.delay(() => {
					assert.equal(wrapper.find('.lucid-SplitVertical.lucid-SplitVertical-is-animated').length, 1);
					done();
				}, 10)
			});

			it('should apply the &-is-animated class when false', (done) => {
				wrapper = mount(
					<SplitVertical isAnimated={false} />
				);

				_.delay(() => {
					assert.equal(wrapper.find('.lucid-SplitVertical.lucid-SplitVertical-is-animated').length, 0);
					done();
				}, 10)
			});
		});

		describeWithDOM('collapseShift', () => {
			let wrapper;
			let mountTestDiv;

			beforeEach(() => {
				mountTestDiv = document.createElement('div');
				document.body.appendChild(mountTestDiv);
			});

			afterEach(() => {
				if (wrapper) {
					wrapper.unmount();
				}

				if (mountTestDiv) {
					mountTestDiv.parentNode.removeChild(mountTestDiv);
				}
			});

			it('should translated by 64px - width when the right pane is primary', (done) => {
				wrapper = mount(
					<SplitVertical isExpanded={false} collapseShift={64}>
						<SplitVertical.LeftPane />
						<SplitVertical.RightPane isPrimary />
					</SplitVertical>
				, { attachTo: mountTestDiv});

				_.delay(() => {
					const innerDiv = mountTestDiv.querySelector('.lucid-SplitVertical-inner');
					const secondaryPaneDiv = mountTestDiv.querySelector('.lucid-SplitVertical-is-secondary');
					const translatePX = parseInt(innerDiv.style.transform.replace(/(translateX[(]|px[)])/ig, ''), 10);
					const width = secondaryPaneDiv.getBoundingClientRect().width;

					assert.equal(64 - width, translatePX, 'must be translated by 64px - width');
					done();
				}, 10)
			});

			it('should translated by width - 64px when the left pane is primary', (done) => {
				wrapper = mount(
					<SplitVertical isExpanded={false} collapseShift={64}>
						<SplitVertical.LeftPane isPrimary />
						<SplitVertical.RightPane />
					</SplitVertical>
				, { attachTo: mountTestDiv});

				_.delay(() => {
					const innerDiv = mountTestDiv.querySelector('.lucid-SplitVertical-inner');
					const secondaryPaneDiv = mountTestDiv.querySelector('.lucid-SplitVertical-is-secondary');
					const translatePX = parseInt(innerDiv.style.transform.replace(/(translateX[(]|px[)])/ig, ''), 10);
					const width = secondaryPaneDiv.getBoundingClientRect().width;

					assert.equal(width - 64, translatePX, 'must be translated by width - 64px');
					done();
				}, 10)
			});

		});

		describeWithDOM('onResize', () => {
			let wrapper;
			let mountTestDiv;

			beforeEach(() => {
				mountTestDiv = document.createElement('div');
				document.body.appendChild(mountTestDiv);
			});

			afterEach(() => {
				if (wrapper) {
					wrapper.unmount();
				}

				if (mountTestDiv) {
					mountTestDiv.parentNode.removeChild(mountTestDiv);
				}
			});

			it('should be called when the DragCaptureZone calls the onDragEnd event handler', () => {
				const onResize = sinon.spy();

				wrapper = mount(
					<SplitVertical isExpanded={true} onResize={onResize} />
				, { attachTo: mountTestDiv });

				const {
					onDragStart,
					onDrag,
					onDragEnd,
				} = wrapper.find(DragCaptureZone).props();

				const lastArg = { event: {} };

				onDragStart(lastArg);
				onDrag({dX: 123}, lastArg);
				onDragEnd({dX: 123}, lastArg);

				assert(onResize.called, 'must be called');
				assert.equal(onResize.lastCall.args[0], 123, 'must pass the new width of the pane');
				assert.equal(onResize.lastCall.args[1].props, wrapper.props(), 'must pass component props in the last arg');
				assert.equal(onResize.lastCall.args[1].event, lastArg.event, 'must pass event reference in the last arg');
			});
		});
	});

	describe('child components', () => {
		describe('LeftPane', () => {
			it('should render children passed in', () => {
				const wrapper = shallow(
					<SplitVertical>
						<SplitVertical.LeftPane>Search Filters</SplitVertical.LeftPane>
					</SplitVertical>
				);

				const LeftPane = wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-LeftPane');

				assert.equal('Search Filters', LeftPane.text(), 'must render children passed in');
			});

			it('should set the right pane as secondary when the left pane is set to primary', () => {
				const wrapper = shallow(
					<SplitVertical>
						<SplitVertical.LeftPane isPrimary>Search Filters</SplitVertical.LeftPane>
					</SplitVertical>
				);

				const RightPane = wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-RightPane');

				assert(RightPane.hasClass('lucid-SplitVertical-is-secondary'), 'must have the secondary className');
			});

			it('should pass thru the with to flexBasis', () => {
				const wrapper = shallow(
					<SplitVertical>
						<SplitVertical.LeftPane width={123}>Search Filters</SplitVertical.LeftPane>
					</SplitVertical>
				);

				const LeftPane = wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-LeftPane');

				assert.equal(123, LeftPane.prop('style').flexBasis, 'must set the flexBasis to match the given width');
			});
		});

		describe('RightPane', () => {
			it('should render children passed in', () => {
				const wrapper = shallow(
					<SplitVertical>
						<SplitVertical.RightPane>Search Filters</SplitVertical.RightPane>
					</SplitVertical>
				);

				const RightPane = wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-RightPane');

				assert.equal('Search Filters', RightPane.text(), 'must render children passed in');
			});

			it('should set the left pane as secondary when the right pane is set to primary', () => {
				const wrapper = shallow(
					<SplitVertical>
						<SplitVertical.RightPane isPrimary>Search Filters</SplitVertical.RightPane>
					</SplitVertical>
				);

				const LeftPane = wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-LeftPane');

				assert(LeftPane.hasClass('lucid-SplitVertical-is-secondary'), 'must have the secondary className');
			});

			it('should pass thru the with to flexBasis', () => {
				const wrapper = shallow(
					<SplitVertical>
						<SplitVertical.RightPane width={123}>Search Filters</SplitVertical.RightPane>
					</SplitVertical>
				);

				const RightPane = wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-RightPane');

				assert.equal(123, RightPane.prop('style').flexBasis, 'must set the flexBasis to match the given width');
			});
		});

		describe('Divider', () => {
			it('should render children passed in', () => {
				const wrapper = shallow(
					<SplitVertical>
						<SplitVertical.Divider>Resize</SplitVertical.Divider>
					</SplitVertical>
				);

				const dividerWrapper = wrapper.find('.lucid-SplitVertical > .lucid-SplitVertical-inner > .lucid-SplitVertical-Divider');

				assert.equal('Resize', dividerWrapper.children().text(), 'must render children passed in');
			});
		});
	});
});
