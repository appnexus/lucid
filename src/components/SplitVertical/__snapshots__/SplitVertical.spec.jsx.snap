// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitVertical [common] example testing should match snapshot(s) for 1.default-split 1`] = `
<div
  className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <Motion
    defaultStyle={
      Object {
        "slideAmount": 0,
      }
    }
    style={
      Object {
        "slideAmount": 0,
      }
    }
  >
    <Component />
  </Motion>
</div>
`;

exports[`SplitVertical [common] example testing should match snapshot(s) for 2.animated-collapse 1`] = `
<div
  className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <Motion
    defaultStyle={
      Object {
        "slideAmount": 0,
      }
    }
    style={
      Object {
        "slideAmount": 0,
      }
    }
  >
    <Component />
  </Motion>
</div>
`;

exports[`SplitVertical [common] example testing should match snapshot(s) for 3.set-pane-as-primary 1`] = `
<div
  className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <Motion
    defaultStyle={
      Object {
        "slideAmount": 0,
      }
    }
    style={
      Object {
        "slideAmount": 0,
      }
    }
  >
    <Component />
  </Motion>
</div>
`;

exports[`SplitVertical [common] example testing should match snapshot(s) for 4.set-width-of-pane 1`] = `
<div
  className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <Motion
    defaultStyle={
      Object {
        "slideAmount": 0,
      }
    }
    style={
      Object {
        "slideAmount": 0,
      }
    }
  >
    <Component />
  </Motion>
</div>
`;

exports[`SplitVertical [common] example testing should match snapshot(s) for 5.customize-divider 1`] = `
<div
  className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <Motion
    defaultStyle={
      Object {
        "slideAmount": 0,
      }
    }
    style={
      Object {
        "slideAmount": 0,
      }
    }
  >
    <Component />
  </Motion>
</div>
`;

exports[`SplitVertical [common] example testing should match snapshot(s) for 6.no-animation 1`] = `
<div
  className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <Motion
    defaultStyle={
      Object {
        "slideAmount": 0,
      }
    }
    style={
      Object {
        "slideAmount": 0,
      }
    }
  >
    <Component />
  </Motion>
</div>
`;

exports[`SplitVertical [common] example testing should match snapshot(s) for 7.collapse-shift 1`] = `
<div
  className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <Motion
    defaultStyle={
      Object {
        "slideAmount": 0,
      }
    }
    style={
      Object {
        "slideAmount": 0,
      }
    }
  >
    <Component />
  </Motion>
</div>
`;

exports[`SplitVertical [common] example testing should match snapshot(s) for 8.handle-resize 1`] = `
<div
  className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <Motion
    defaultStyle={
      Object {
        "slideAmount": 0,
      }
    }
    style={
      Object {
        "slideAmount": 0,
      }
    }
  >
    <Component />
  </Motion>
</div>
`;

exports[`SplitVertical [common] example testing should match snapshot(s) for 9.no-resize 1`] = `
<div
  className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <Motion
    defaultStyle={
      Object {
        "slideAmount": 0,
      }
    }
    style={
      Object {
        "slideAmount": 0,
      }
    }
  >
    <Component />
  </Motion>
</div>
`;

exports[`SplitVertical render should render expanded & collapsed 1`] = `
<SplitVertical
  collapseShift={0}
  isAnimated={false}
  isExpanded={true}
  isResizeable={true}
  onResize={[Function]}
  onResizing={[Function]}
>
  <div
    className="lucid-SplitVertical lucid-SplitVertical-is-expanded"
    style={
      Object {
        "overflow": "hidden",
      }
    }
  >
    <Motion
      defaultStyle={
        Object {
          "slideAmount": 0,
        }
      }
      style={
        Object {
          "slideAmount": 0,
        }
      }
    >
      <div
        className="lucid-SplitVertical-inner"
        style={
          Object {
            "display": "flex",
            "transform": "translateX(0px)",
          }
        }
      >
        <div
          className="lucid-SplitVertical-LeftPane lucid-SplitVertical-is-secondary"
          style={
            Object {
              "flexBasis": "calc(50% - 3px)",
              "flexGrow": 0,
              "flexShrink": 0,
              "marginLeft": undefined,
              "overflow": "auto",
            }
          }
        />
        <DragCaptureZone
          className="lucid-SplitVertical-Divider lucid-SplitVertical-Divider-is-resizeable"
          onDrag={[Function]}
          onDragCancel={[Function]}
          onDragEnd={[Function]}
          onDragStart={[Function]}
          style={
            Object {
              "boxSizing": "border-box",
              "width": "6px",
            }
          }
        >
          <div
            className="lucid-DragCaptureZone lucid-SplitVertical-Divider lucid-SplitVertical-Divider-is-resizeable"
            key="DragCaptureZone"
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "boxSizing": "border-box",
                "width": "6px",
              }
            }
          >
             
          </div>
        </DragCaptureZone>
        <div
          className="lucid-SplitVertical-RightPane"
          style={
            Object {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
              "marginRight": -0,
              "overflow": "auto",
            }
          }
        />
      </div>
    </Motion>
  </div>
</SplitVertical>
`;

exports[`SplitVertical render should render expanded & collapsed 2`] = `
<div
  class="lucid-SplitVertical"
  style="overflow: hidden;"
>
  <div
    class="lucid-SplitVertical-inner"
    style="display: flex; transform: translateX(0px);"
  >
    <div
      class="lucid-SplitVertical-LeftPane lucid-SplitVertical-is-secondary"
      style="flex-grow: 0; flex-shrink: 0; flex-basis: calc(50% - 3px); overflow: auto;"
    />
    <div
      class="lucid-DragCaptureZone lucid-SplitVertical-Divider lucid-SplitVertical-Divider-is-resizeable"
      style="width: 6px; box-sizing: border-box;"
    >
       
    </div>
    <div
      class="lucid-SplitVertical-RightPane"
      style="flex-grow: 1; flex-shrink: 1; flex-basis: 0%; margin-right: 0px; overflow: auto;"
    />
  </div>
</div>
`;
